src/Core/src/Hosting/ImageSources/ImageSourceToImageSourceServiceTypeMapping.cs

```diff
@@ -0,0 +1,111 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+using TypePair = (System.Type ImageSource, System.Type ImageSourceService);
+
+namespace Microsoft.Maui.Hosting
+{
+	internal sealed class ImageSourceToImageSourceServiceTypeMapping
+	{
+		private static readonly ConcurrentDictionary<IImageSourceServiceCollection, ImageSourceToImageSourceServiceTypeMapping> s_instances = new();
+
+		internal static ImageSourceToImageSourceServiceTypeMapping GetInstance(IImageSourceServiceCollection collection) =>
+			s_instances.GetOrAdd(collection, static _ => new ImageSourceToImageSourceServiceTypeMapping());
+
+		private ConcurrentDictionary<Type, Type> _interfaceMapping { get; } = new();
+		private ConcurrentDictionary<Type, Type> _directServiceMapping { get; } = new();
+		private ConcurrentDictionary<Type, Type> _fallbackServiceMapping { get; } = new();
+
+		public void Add<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)] TImageSource>() where TImageSource : IImageSource
+		{
+			// override any previous mapping (possibly a default mapping)
+			_directServiceMapping[typeof(TImageSource)] = typeof(IImageSourceService<TImageSource>);
+
+			var imageSourceInterfaceType = GetImageSourceInterface(typeof(TImageSource));
+			_interfaceMapping[typeof(TImageSource)] = imageSourceInterfaceType;
+
+			// add a fallback value for the interface to mimic the behavior of the old implementation
+			AddFallbackMapping();
+
+			[UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+				Justification = "The type which is used as a generic parameter is not a value type.")]
+			void AddFallbackMapping()
+			{
+				if (imageSourceInterfaceType.IsValueType)
+				{
+					throw new InvalidOperationException($"Unable to register {typeof(TImageSource)} as an {typeof(IImageSource)} because it is a value type.");
+				}
+
+				var imageSourceServiceType = typeof(IImageSourceService<>).MakeGenericType(imageSourceInterfaceType);
```

Comment:
@vitek-karas do I remember correctly that using `MakeGenericType` in this context (the generic argument is a reference type) is OK?

Reply:
Well, I'm not 100% about interfaces, but I think it should be OK.
Adding @MichalStrehovsky to confirm.

Also adding @sbomer as this is one of the areas we wanted to improve analysis for - ideally this would not produce a warning at all, but that's not implemented yet in the analyzer/AOT.

Reply:
Sorry - I didn't look at this right - I think this is correct and will not cause problems in AOT.

Reply:
Just curious - what is this `Type` used for? It doesn't have any dataflow annotations, so we can't invoke any methods on it. And the interface doesn't even appear to have any methods in the first place.

I tried to trace the source code through github UI but I couldn't find the "consumer".

Reply:
The type is used just for a service descriptor in DI.

I'm currently thinking that this might be an overkill anyway and we could probably dial it back down to one of the earlier commits which correctly resolves the image source service type but breaks the behavior of `IImageSourceServiceProvider.GetImageSourceServiceType` and `IImageSourceServiceProvider.GetImageSourceType` methods.

We don't need these methods in MAUI internally anywhere and we should discourage customers from using it directly (there doesn't seem to be an obvious good use case for using it anyway).